hazelcast/src/main/java/com/hazelcast/cp/CPSubsystemManagementService.java

```diff
@@ -0,0 +1,215 @@
+     * remaining members of the given CP group. It also performs a Raft commit
+     * to the Metadata CP group to update status of the destroyed group.
+     * Once a CP group id is destroyed, all CP data structure proxies created
+     * before the destroy will fail with {@link CPGroupDestroyedException}.
+     * <p>
+     * Once a CP group is destroyed, it can be created again with a new set of
+     * CP members.
+     * <p>
+     * This method is idempotent. It has no effect if the given CP group is
+     * already destroyed.
+     */
+    ICompletableFuture<Void> forceDestroyCPGroup(String groupName);
+
+    /**
+     * Returns the current list of CP members
+     */
+    ICompletableFuture<Collection<CPMember>> getCPMembers();
+
+    /**
+     * Promotes the local Hazelcast member to a CP member.
+     * <p>
+     * This method is idempotent.
+     * If the local member is already in the active CP members list, then this
+     * method will have no effect. When the current member is promoted to CP
+     * member, its member UUID is assigned as CP member UUID.
+     * <p>
+     * Once the returned {@code Future} object is completed, the promoted CP
+     * member has been added to the CP groups that have missing members, i.e.,
+     * whose size is smaller than {@link CPSubsystemConfig#getGroupSize()}.
+     * <p>
+     * If the local member is currently being removed from
+     * the active CP members list, then the returned {@code Future} object
+     * will throw {@link IllegalArgumentException}.
+     * <p>
+     * If there is an ongoing membership change in the CP subsystem when this
+     * method is invoked, then the returned {@code Future} object will throw
+     * {@link IllegalStateException}
+     *
+     * @throws IllegalArgumentException If the local member is currently being
+     *         removed from the active CP members list
+     * @throws IllegalStateException If there is an ongoing membership change
+     *         in the CP subsystem
+     */
+    ICompletableFuture<Void> promoteToCPMember();
+
+    /**
+     * Removes the given unreachable CP member from the active CP members list
+     * and all CP groups it belongs to. If any other active CP member
+     * is available, it will replace the removed CP member in its CP groups.
+     * Otherwise, CP groups which the removed CP member is a member of will
+     * shrink and their majority values will be recalculated.
+     * <p>
+     * Before removing a CP member from the CP subsystem, please make sure that
+     * it is declared as unreachable by Hazelcast's failure detector and removed
+     * from Hazelcast's member list. The behaviour is undefined when a running
+     * CP member is removed from the CP subsystem.
+     *
+     * @throws IllegalStateException When another CP member is being removed
+     *         from the CP sub-system
+     * @throws IllegalArgumentException if the given CP member is still part
```

Comment:
I don't quite understand this part. So we can't remove a member which is currently a part of the cluster (e.g. degrade it)? If we want to remove it, we must shut it down?

And I can't seem to find how where this exception is thrown. I only see a log message [here](https://github.com/hazelcast/hazelcast/blob/d7c7d2b58b07f4204ffcdb36bdf04cc7ac4e983c/hazelcast/src/main/java/com/hazelcast/cp/internal/RaftService.java#L368-L369)

Reply:
`@throws IllegalArgumentException` is a leftover. It was being thrown before but we had to remove that part and replaced with warning log.

Reply:
Actually, there's no need to remove a CP member if it's already alive. Removal is needed when a CP member is crashed. We cannot remove automatically, since we cannot distinguish whether it's crashed or unreachable.

Reply:
Yes, I was thinking more if there was a way to "demote" a CP member to a vanilla HZ member.